energymeters/integrationpluginenergymeters.cpp

```diff
@@ -0,0 +1,394 @@
+
+    m_powerFactorStateTypeIds.insert(pro380ThingClassId, pro380PowerFactorStateTypeId);
+    m_powerFactorStateTypeIds.insert(sdm630ThingClassId, sdm630PowerFactorStateTypeId);
+
+    m_totalEnergyConsumedStateTypeIds.insert(pro380ThingClassId, pro380TotalEnergyConsumedEventTypeId);
+    m_totalEnergyConsumedStateTypeIds.insert(sdm630ThingClassId, sdm630TotalEnergyConsumedEventTypeId);
+
+    m_totalEnergyProducedStateTypeIds.insert(pro380ThingClassId, pro380TotalEnergyProducedStateTypeId);
+    m_totalEnergyProducedStateTypeIds.insert(sdm630ThingClassId, sdm630TotalEnergyProducedStateTypeId);
+
+    m_discoverySlaveAddressParamTypeIds.insert(pro380ThingClassId, pro380DiscoverySlaveAddressParamTypeId);
+    m_discoverySlaveAddressParamTypeIds.insert(sdm630ThingClassId, sdm630DiscoverySlaveAddressParamTypeId);
+
+    m_registerMaps.insert(pro380ThingClassId, pro380RegisterMap);
+    m_registerMaps.insert(sdm630ThingClassId, sdm630RegisterMap);
+
+    // FIXME leads to crash, probably an issue in the modbus rtu resource
+    /*connect(hardwareManager()->modbusRtuResource(), &ModbusRtuHardwareResource::modbusRtuMasterRemoved, this, [=] (const QUuid &modbusUuid){
+        qCDebug(dcEnergyMeters()) << "Modbus RTU master has been removed" << modbusUuid.toString();
+
+        // Check if there is any device using this resource
+        foreach (EnergyMeter * meter, m_energyMeters) {
+            if (meter->modbusRtuMasterUuid() == modbusUuid) {
+                Thing *thing = m_energyMeters.key(meter);
+                if (!thing)
+                    return;
+
+                qCWarning(dcEnergyMeters()) << "Modbus RTU hardware resource removed for" << thing << ". The thing will not be functional any more until a new resource has been configured for it.";
+                thing->setStateValue(m_connectionStateTypeIds[thing->thingClassId()], false);
+            }
+        }
+    }); */
+}
+
+void IntegrationPluginEnergyMeters::discoverThings(ThingDiscoveryInfo *info)
+{
+    qCDebug(dcEnergyMeters()) << "Discover things";
+    QList<ThingDescriptor> thingDescriptors;
+
+    if (hardwareManager()->modbusRtuResource()->modbusRtuMasters().isEmpty()) {
+        info->finish(Thing::ThingErrorHardwareNotAvailable, QT_TR_NOOP("No Modbus RTU interface available."));
+        return;
+    }
+
+    if (m_connectionStateTypeIds.contains(info->thingClassId())) {
+        int slaveAddress = info->params().paramValue(m_discoverySlaveAddressParamTypeIds.value(info->thingClassId())).toInt();
+        if (slaveAddress > 254 || slaveAddress == 0) {
+            info->finish(Thing::ThingErrorInvalidParameter, tr("Modbus slave address must be between 1 and 254"));
+            return;
+        }
+        Q_FOREACH(ModbusRtuMaster *modbusMaster, hardwareManager()->modbusRtuResource()->modbusRtuMasters()) {
+            qCDebug(dcEnergyMeters()) << "Found RTU master resource" << modbusMaster << "connected" << modbusMaster->connected();
```

Comment:
Possible improvement: if you know that the meter requires a certain baudrate, databits, stopbits etc. you could filter here for the correct configured resources by checking the configuration of each master. Only return the compatible one. Not sure if that makes sense here, just a thought :-)

Reply:
Also thought about this, but all the energy meters I tested had the possibility to change the modbus settings